### PR TITLE
Make inreplace a purely static method v2

### PR DESCRIPTION
**Pull request by dduugg**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
Revised attempt at https://github.com/Homebrew/brew/pull/15807, this time including tests and explicit block forwarding

cc @bayandin

### CODE REVIEW

File: Library/Homebrew/test/dev-cmd/test_spec.rb [hunk 1]
```diff
@@ -17,4 +17,40 @@
       .and not_to_output.to_stderr
       .and be_a_success
   end
+
+  describe "using inreplace" do
```
Comment (apainintheneck) [inline Library/Homebrew/test/dev-cmd/test_spec.rb:21]: We usually try to limit the number of integration tests if possible. Is there any way to test this without integration tests?

Comment (dduugg) [reply Library/Homebrew/test/dev-cmd/test_spec.rb:21]: Probably! How does one get a generic formula instance in a unit test? (`inreplace` could be a static method, it doesn't use `Formula` state).

(Though I do think we should have more integration tests using formula `test` blocks, such as the one used to repro this issue: https://github.com/Homebrew/brew/pull/15807#issuecomment-1664172239 )

Comment (apainintheneck) [reply Library/Homebrew/test/dev-cmd/test_spec.rb:21]: Ideally, we'd add some tests like those found in the `formula_spec.rb` file. Here's an example with the livecheck block.

https://github.com/Homebrew/brew/blob/3eb48a828d44643b8d427f647227a277b1a4c1bf/Library/Homebrew/test/formula_spec.rb#L665-L679

If that's too much of an ask, maybe we could just combine these integration tests. I think that would still result in less CI time.

Comment (apainintheneck) [reply Library/Homebrew/test/dev-cmd/test_spec.rb:21]: We don't seem to test the test blocks that much though.

File: Library/Homebrew/formula.rb [hunk 2]
```diff
@@ -2560,10 +2559,11 @@ def install; end
       before:       T.nilable(T.any(Pathname, Regexp, String)),
       after:        T.nilable(T.any(Pathname, String, Symbol)),
       audit_result: T::Boolean,
+      blk:          T.nilable(T.proc.params(s: StringInreplaceExtension).void),
     ).void
   }
-  def inreplace(paths, before = nil, after = nil, audit_result = true) # rubocop:disable Style/OptionalBooleanParameter
-    super(paths, before, after, audit_result)
+  def inreplace(paths, before = nil, after = nil, audit_result = true, &blk) # rubocop:disable Style/OptionalBooleanParameter
+    Utils::Inreplace.inreplace(paths, before, after, audit_result: audit_result, &blk)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb:2566]:
```suggestion
  def inreplace(paths, before = nil, after = nil, audit_result = true, &block) # rubocop:disable Style/OptionalBooleanParameter
    Utils::Inreplace.inreplace(paths, before, after, audit_result: audit_result, & block)
```

File: Library/Homebrew/test/dev-cmd/test_spec.rb (hunk 1 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/test/dev-cmd/test_spec.rb:21]: Agreed these would ideally not be integration tests.

File: Library/Homebrew/utils/inreplace.rb [hunk 3]
```diff
@@ -43,9 +41,10 @@ def initialize(errors)
         before:       T.nilable(T.any(Pathname, Regexp, String)),
         after:        T.nilable(T.any(Pathname, String, Symbol)),
         audit_result: T::Boolean,
+        blk:          T.nilable(T.proc.params(s: StringInreplaceExtension).void),
       ).void
     }
-    def inreplace(paths, before = nil, after = nil, audit_result = true) # rubocop:disable Style/OptionalBooleanParameter
+    def self.inreplace(paths, before = nil, after = nil, audit_result: true, &blk)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/inreplace.rb:47]:
```suggestion
    def self.inreplace(paths, before = nil, after = nil, audit_result: true, &block)
```

File: Library/Homebrew/utils/inreplace.rb [hunk 4]
```diff
@@ -43,9 +41,10 @@ def initialize(errors)
         before:       T.nilable(T.any(Pathname, Regexp, String)),
         after:        T.nilable(T.any(Pathname, String, Symbol)),
         audit_result: T::Boolean,
+        blk:          T.nilable(T.proc.params(s: StringInreplaceExtension).void),
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/inreplace.rb:44]:
```suggestion
        block:          T.nilable(T.proc.params(s: StringInreplaceExtension).void),
```

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks again @dduugg!

**Pull request merged**